src/main/java/com/mlaf/hu/decisionagent/DecisionAgent.java

```diff
@@ -0,0 +1,102 @@
+import com.mlaf.hu.decisionagent.behavior.RegisterBehavior;
+import com.mlaf.hu.decisionagent.behavior.UpdateBehavior;
+import com.mlaf.hu.helpers.JadeServices;
+import com.mlaf.hu.models.*;
+import jade.core.AID;
+import jade.core.Agent;
+import jade.domain.DFService;
+import jade.util.Logger;
+
+import javax.xml.bind.JAXB;
+import java.io.StringReader;
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.HashMap;
+
+public abstract class DecisionAgent extends Agent {
+    private static final String SERVICE_NAME = "DECISION-AGENT";
+    private static final int MAX_READINGS = 100;
+    public static java.util.logging.Logger decisionAgentLogger = Logger.getLogger("DecisionAgentLogger");
+    public HashMap<AID, InstructionSet> sensorAgents = new HashMap<>();
+
+    public DecisionAgent() {
+        super();
+    }
+
+    @Override
+    protected void setup() {
+        try {
+            JadeServices.registerAsService(SERVICE_NAME, "decision-agent", null, null, this);
+            addBehaviour(new RegisterBehavior(this));
+            addBehaviour(new ReceiveBehavior(this));
+            addBehaviour(new UpdateBehavior(this));
+        } catch (Exception e) {
+            DecisionAgent.decisionAgentLogger.log(Logger.SEVERE, "Could not initialize BrokerAgent", e);
+            System.exit(1);
+        }
+    }
+
+    protected void takeDown() {
+        try {
+            DFService.deregister(this);
+        } catch (Exception ignore) {
+        }
+    }
+
+    public void registerSensorAgent(AID sensoragent, InstructionSet instructionset) {
+        this.sensorAgents.put(sensoragent, instructionset);
+    }
+
+    public abstract void unregisterSensorAgent(AID sensoragent);
+
+    public InstructionSet parseInstructionXml(String xml) {
+        InstructionSet is = new InstructionSet();
+        try {
+            is = JAXB.unmarshal(new StringReader(xml), InstructionSet.class);
+        } catch (Exception e) {
+            decisionAgentLogger.log(Logger.SEVERE, String.format("Error parsing XML: %s", e.getMessage()));
+        }
+        return is;
+    }
+
+    public SensorReading parseSensorReadingXml(String xml) {
+        SensorReading sr = new SensorReading();
+        try {
+            sr = JAXB.unmarshal(new StringReader(xml), SensorReading.class);
+        } catch (Exception e) {
+            decisionAgentLogger.log(Logger.SEVERE, String.format("Error parsing XML: %s", e.getMessage()));
+        }
+        return sr;
+    }
+
+    public void handleSensorReading(double value, InstructionSet is, Sensor sensor, String measurementId) {
+        try {
+            Measurement measurement = sensor.getMeasurements().getMeasurement(measurementId);
+            ArrayList<Double> readings = measurement.getReadings();
+            readings.add(value);
+            is.setLastReceivedDataPackageAt(LocalDateTime.now());
+            if (readings.size() > MAX_READINGS) {
```

Comment:
Misschien beter om voor de readings een CircularFifoQueue uit Apache Commons te gebruiken.
Als je deze aanmaakt geef je een max size aan, zodra hij hier overheen gaat en je iets nieuws wil inserten, gooit ie de oudste eerst weg.

http://commons.apache.org/proper/commons-collections/javadocs/api-release/org/apache/commons/collections4/queue/CircularFifoQueue.html

Reply:
Perfect! Ik zat even te googlen, maar kon dat dus niet vinden. Ga ik implementeren.

Reply:
@aacoba is dat ook iets dat gebruikt zou kunnen worden in plaats van mijn Cache klasse voor die monitoringservice? (die je overigens al goed gekeurd hebt)

Reply:
@rogierpijpers Ja daar zat ik net ook aan te denken, is misschien wel net zo netjes voor consistency, zeker als we dan toch al apache commons gaan gebruiken kan het geen kwaad.

Reply:
Geïmplementeerd.